### PR TITLE
Add HintColor, SetCursorAt and SetCursorAtEnd for TextBoxes

### PR DESCRIPTION
**Pull request by gordonlu310**
As part of the App Inventor + Kodular collaboration, I have prepared the following blocks and sorted them into a PR.

- **Add HintColor property** to TextBox, PasswordTextBox and EmailPicker.
- **Add SetCursorAt and SetCursorAtEnd methods** to TextBox, PasswordTextBox and EmailPicker.

This PR will also fix the bug pointed out here: [FontTypeface of PasswordTextBox is reset after toggling password visibility](https://github.com/mit-cml/appinventor-sources/issues/2966#issue-1871340468).

I cannot program in Swift, so this PR does not include iOS yet. As I am still a newbie in open source development I may make mistakes. Please point them out, and I'm happy to correct them!

### CODE REVIEW

**Comment by AppInventorWorkerBee** [comment]
Can one of the admins verify this patch?

**Comment by ewpatton** [comment]
@gordonlu310 This looks pretty good at first glance. However, any changes that modify the components by adding new blocks should have `ucr` as the base branch for the PR rather than `master`. You should rebase these changes onto `ucr` and update this PR's base branch accordingly.

**Comment by gordonlu310** [comment]
@ewpatton I just edited the base branch in GitHub, and a lot of other files are changed (including ProGuard)! What do I do?

**Comment by ewpatton** [comment]
Assuming you have local branches master and ucr and they are up-to-date with ours, you should be able to do the following from your `patch-5` branch:

    git rebase master --onto ucr

If there are errors, you will need to resolve the conflicts. After completing the rebase, you will need to force-push with `git push -f` to update your branch on GitHub since its history will have changed.

**Comment by gordonlu310** [comment]
@ewpatton Thanks, I believe it's completed now.

File: appinventor/components/src/com/google/appinventor/components/runtime/TextBoxBase.java [hunk 1]
```diff
@@ -432,6 +436,28 @@ public void Hint(String hint) {
...
+
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_COLOR, defaultValue = Component.DEFAULT_VALUE_COLOR_GRAY)
+  @SimpleProperty()
+  public void HintColor(int hintColor) {
+    this.hintColor = hintColor;
+    if (hintColor != 0xFF888888) {
+      view.setHintTextColor(hintColor);
+    } else {
+      if (isHighContrast || container.$form().HighContrast()) {
+        view.setHintTextColor(COLOR_YELLOW);
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/TextBoxBase.java:454]: Does yellow work in both light and dark themes?

**Review by ewpatton — CHANGES_REQUESTED** [review]
A few small comments.

File: appinventor/components/src/com/google/appinventor/components/runtime/TextBoxBase.java [hunk 2]
```diff
@@ -48,17 +49,20 @@ public abstract class TextBoxBase extends AndroidViewComponent
...
-  private String fontTypeface;
+  protected String fontTypeface;
 
   // Backing for font bold
-  private boolean bold;
+  protected boolean bold;
 
   // Backing for font italic
-  private boolean italic;
+  protected boolean italic;
```
[the remark this answers is not in the document]
Comment (gordonlu310) [reply appinventor/components/src/com/google/appinventor/components/runtime/TextBoxBase.java:58]: Thanks. I thought there is no getter block for bold, italic and font typeface. Will change it.

File: appinventor/components/src/com/google/appinventor/components/runtime/TextBoxBase.java (hunk 1 above)
Comment (gordonlu310) [reply appinventor/components/src/com/google/appinventor/components/runtime/TextBoxBase.java:454]: Yes.

File: appinventor/components/src/com/google/appinventor/components/runtime/PasswordTextBox.java [hunk 3]
```diff
@@ -81,6 +82,7 @@ public PasswordTextBox(ComponentContainer container) {
   public void PasswordVisible(boolean visible){
     passwordVisible=visible;
     setPasswordInputType(acceptsNumbersOnly, visible);
+    TextViewUtil.setFontTypeface(container.$form(), view, fontTypeface, bold, italic);
```
[the remark this answers is not in the document]
Comment (gordonlu310) [reply appinventor/components/src/com/google/appinventor/components/runtime/PasswordTextBox.java]: Noted. Changed now.

File: appinventor/components/src/com/google/appinventor/components/runtime/TextBoxBase.java [hunk 4]
```diff
@@ -48,17 +49,20 @@ public abstract class TextBoxBase extends AndroidViewComponent
...
 
   // Backing for hint text
   private String hint;
 
+  // Backing for hint color
+  private int hintColor = 0xFF888888;
```
[the remark this answers is not in the document]
Comment (gordonlu310) [reply appinventor/components/src/com/google/appinventor/components/runtime/TextBoxBase.java]: Noted. Changed now.

**Comment by ewpatton** [comment]
@AppInventorWorkerBee ok to test

**Comment by gordonlu310** [comment]
@ewpatton Are the changes OK now?

**Review by ewpatton — APPROVED** [review]
LGTM

**Review by ewpatton — CHANGES_REQUESTED** [review]
Having thought about it a little bit, I'd also like to see a `SetCursorAtStart` block to provide symmetry with `SetCursorAtEnd`. This would also be helpful for new users who may not yet understand 0 vs 1 based indexing, which the cursor manipulation relies on, so `SetCursorAtStart` makes the semantics explicit. If you could add that one block and rebuild the documentation this should be good to go.

**Comment by ewpatton** [comment]
@gordonlu310 Also, after thinking about it I think that the SetCursorAt() function should be 1-indexed rather than 0-indexed, so we will need to subtract 1 from the incoming index. This will make use of this function compatible with the text blocks without requiring the user to do additional computations. For example:

blocks (83)

If in the TextBox1 you put "abcdefg" and then search for "cde" the cursor will end up between the "c" and "d", whereas it seems like it should show up at the start of "cde".

**Comment by ewpatton** [comment]
We probably also want to clamp the value to the valid range. Anything less than 1 effectively moves the cursor to the beginning and anything greater than the length of the string moves to the end. I have not tried yet, for example, passing -1 and seeing what happens, but it must not crash the program.

**Comment by ewpatton** [comment]
@gordonlu310 After discussing with the team, along with the change to indexing we think that a name of `SetCursorBefore` would be more clear than `SetCursorAt`. I'm also thinking that "Move" rather than "Set" would be more clear to beginners, but the jury is still out on that one.

**Comment by gordonlu310** [comment]
> @gordonlu310 After discussing with the team, along with the change to indexing we think that a name of `SetCursorBefore` would be more clear than `SetCursorAt`. I'm also thinking that "Move" rather than "Set" would be more clear to beginners, but the jury is still out on that one.

I have also thought about this before posting the PR, but it would be a bit hard for Kodular folks to migrate their projects containing the blocks here. We will have to deprecate the blocks and describe the situation in the documentation, which could be a bit confusing.

**Comment by ewpatton** [comment]
We will be creating a migration tool to handle these types of situations, so I would not worry about that. We will need to keep a log of what needs to be migrated, and we have code in App Inventor already to perform method renames. Sent from my iPhoneOn Oct 16, 2023, at 19:37, Gordon Lu ***@***.***> wrote:﻿

@gordonlu310 After discussing with the team, along with the change to indexing we think that a name of SetCursorBefore would be more clear than SetCursorAt. I'm also thinking that "Move" rather than "Set" would be more clear to beginners, but the jury is still out on that one.

I have also thought about this before posting the PR, but it would be a bit hard for Kodular folks to migrate their projects containing the blocks here. We will have to deprecate the blocks and describe the situation in the documentation, which could be a bit confusing.

**Comment by ewpatton** [comment]
Yes, you can see how we did this for the Spreadsheet component in versioning.js:

https://github.com/mit-cml/appinventor-sources/blob/17b754f9cafae1246a7cdd564189212843de46a6/appinventor/blocklyeditor/src/versioning.js#L3086

However, we'd want to wait until we are ready for the Kodular transition since we will need to bump the version numbers for all affected components so that the versioning code will run on first import. If our version numbers are still behind the Kodular ones, nothing will happen and the project will appear corrupted.

**Comment by gordonlu310** [comment]
Sure, I'll leave it here (for now) and work on other issues.

**Comment by gordonlu310** [comment]
@ewpatton When I create a copy of my local ucr branch using `git checkout -b new_branch ucr`, add/modify the files and checkout to the ucr branch, the changes are applied in the ucr branch. Running `git checkout -- file` removes the changes from both branches. How do I avoid this?

**Comment by patryk84a** [comment]
If this is to be compatible with Kodular, shouldn't the method naming be identical to Kodular? If the methods have a different name, Kodular projects imported into AppInventor will crash.

**Comment by ewpatton** [comment]
@gordonlu310 You should verify you've actually switched the branch using `git branch`. When you commit, it will be on whatever branch is returned there.

Is this PR ready for a re-review?

**Comment by gordonlu310** [comment]
> @gordonlu310 You should verify you've actually switched the branch using `git branch`. When you commit, it will be on whatever branch is returned there.
> 
> Is this PR ready for a re-review?

Thanks. Not yet, I still have to change a few more things.

**Comment by gordonlu310** [comment]
I'm sorry, I have competitions and exams at the moment, and so I wasn't able to work on the PR. I think I may be able to work on it a few days later. Thanks and sorry!

**Comment by ewpatton** [comment]
@gordonlu310 This was pretty close to being done. I'm happy to take it up and finish it if you do not have the time.

**Comment by gordonlu310** [comment]
No, it's fine. I messed up a lot of things in my repository, so I want to start over again. I've fixed and tested everything, works like butter. I'll post it in a few hours.

**Pull request closed without merging**